Automated Smoketest: :x:
Sha: 88c9283fae506e290a815680fb611a6338a471e0
Build: failed
Number of Error(s): 14
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:105:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:108:1: error: expected unqualified-id before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:110:1: error: expected unqualified-id before \xe2\x80\x98>>\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:90:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:90:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:90:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:90:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:732:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10323/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:1654:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
make[2]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/unicodelib.cpp.o] Error 1
make[1]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

--- a/ecllibrary/std/Uni.ecl
+++ b/ecllibrary/std/Uni.ecl
@@ -449,4 +449,17 @@ EXPORT Translate(unicode text, unicode sear, unicode repl) :=
 EXPORT BOOLEAN StartsWith(unicode src, unicode pref, string form) :=
 	lib_unicodelib.UnicodeLib.UnicodeLocaleStartsWith(src, pref, form);
 
+/**
+ * Returns true if the suffix string matches the trailing characters in the source string.  Trailing and Leading spaces
+ * are stripped from the suffix before matching. Unless specified, normalization will not occur. Unless initiated as hex and
+ * then converted to Unicode using TRANSFER, ecl will perform its own normalization on your declared Unicode string.
+ *
+ * @param src           The string being searched in.
+ * @param suff          The suffix to search for.
+ * @param form          The type of Normalization to be employed.
+ */
+
+EXPORT BOOLEAN EndsWith(unicode src, unicode suff, string form) :=
+	lib_unicodelib.UnicodeLib.UnicodeLocaleEndsWith(src, suff, form);
+
 END;

--- a/ecllibrary/teststd/uni/TestEndsWith.ecl
+++ b/ecllibrary/teststd/uni/TestEndsWith.ecl
@@ -1,0 +1,98 @@
+﻿/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems®.  All rights reserved.
+############################################################################## */
+
+IMPORT Std.Uni;
+
+EXPORT TestEndsWith := MODULE
+
+  EXPORT TestConst := MODULE
+
+    //Check action on strings with no entries: empty source string, search string, or return string.
+    EXPORT Test01 := ASSERT(Uni.EndsWith('','','') = FALSE, CONST);
+    EXPORT Test03 := ASSERT(Uni.EndsWith('x','','') = FALSE, CONST);
+    EXPORT Test05 := ASSERT(Uni.EndsWith('','x','') = FALSE, CONST);
+    EXPORT Test07 := ASSERT(Uni.EndsWith('','','x') = FALSE, CONST);
+    EXPORT Test02 := ASSERT(Uni.EndsWith('x','x','') = TRUE, CONST);
+    //Check action on a string containing a single word - with various whitespace.
+    EXPORT Test08 := ASSERT(Uni.EndsWith(' x ','  x ','') = FALSE);
+    EXPORT Test09 := ASSERT(Uni.EndsWith(' x','x  ','') = TRUE);
+    EXPORT Test10 := ASSERT(Uni.EndsWith('x ','x ','') = FALSE);
+    EXPORT Test11 := ASSERT(Uni.EndsWith('  x','x','') = TRUE);
+    //Functionality Test.
+    EXPORT Test12 := ASSERT(Uni.EndsWith('x','xx','') = FALSE);
+    EXPORT Test14 := ASSERT(Uni.EndsWith('','','yw') = FALSE);
+    //Check action on a string containg multiple words - with various whitespace combinations.
+    EXPORT Test15 := ASSERT(Uni.EndsWith(' xyz abc','xyz abc','') = TRUE);
+    EXPORT Test16 := ASSERT(Uni.EndsWith(' xyz abc','xyz','') = FALSE);
+    EXPORT Test17 := ASSERT(Uni.EndsWith(' xyz abc','c','') = TRUE);
+    EXPORT Test18 := ASSERT(Uni.EndsWith(' xyz abc','bc','') = TRUE);
+    EXPORT Test19 := ASSERT(Uni.EndsWith(' xyz abc','ab','') = FALSE);
+    EXPORT Test20 := ASSERT(Uni.EndsWith(' xyz abc','b c','') = FALSE);
+    //Check action on a string containing punctuation characters.
+    EXPORT Test21 := ASSERT(Uni.EndsWith(',&%$@','$@','') = TRUE);
+    //Check action on a string containing an apostrophe.
+    EXPORT Test23 := ASSERT(Uni.EndsWith('I couldn\'t hear you!','couldn\'t hear you!','') = TRUE);
+    //Check action on a string containing different variations/combinations of numbers and other characters.
+    EXPORT Test24 := ASSERT(Uni.EndsWith('123abc 23.6 abc123','23.6 abc123','') = TRUE);
+    //Test other space characters (< 0x20).
+    EXPORT Test26 := ASSERT(Uni.EndsWith('\nt\t','t','') = FALSE);
+    //Check action on a string containing latin diacritical marks.
+    EXPORT Test27 := ASSERT(Uni.EndsWith(U'À à',U'à','') = TRUE);
+    EXPORT Test28 := ASSERT(Uni.EndsWith(U'ȭ š',U'š','') = TRUE);
+    //Check action on a string containing Spanish words with latin accents.
+    //Translation: "The deceased changed the girls"
+    EXPORT Test29 := ASSERT(Uni.EndsWith(U'El difunto cambió las niñas',U'niñas','') = TRUE);
+    //Check action on a string containing Chinese characters.
+    //Translation: "I am a computer"
+    EXPORT Test30 := ASSERT(Uni.EndsWith(U'我是電腦',U'腦','') = TRUE);
+    //Check action on a string containing Modern Greek characters.
+    //Translation: "Do you come here often?"
+    EXPORT Test31 := ASSERT(Uni.EndsWith(U'Έρχεσαι συχνά εδώ;',U'εδώ;','') = TRUE);
+    //Testcases 32 and 33 test for bidirectional capabilities with scripts in arabic and hebrew.
+    //Check action on arabic lettering with accent marks. Bidirectional.
+    //Translation: "Good morning"
+    EXPORT Test32 := ASSERT(Uni.EndsWith(U'صباح الخير',U'الخير','') = TRUE);
+    //Check action on hebrew lettering with accent marks (called pointing). Bidirectional.
+    //Translation: (not a phrase, 2 different words separated by a space)
+    EXPORT Test33 := ASSERT(Uni.EndsWith(U'קָמָץ שִׁי״ן',U'שִׁי״ן','') = TRUE);
+    //Check action on surrogate pairs.
+    EXPORT Test34 := ASSERT(Uni.EndsWith(U'x𐐀',U'x𐐀','') = TRUE);
+    EXPORT Test35 := ASSERT(Uni.EndsWith(U'𐐀',U'𐐀x','') = FALSE);
+    EXPORT Test37 := ASSERT(Uni.EndsWith(U'x𐐀',U'𐐀','') = TRUE);
+    //Check action with normalization forms
+    EXPORT Test39 := ASSERT(Uni.EndsWith(U'Ç̌',U'Ç̌','') = TRUE);
+    EXPORT Test40 := ASSERT(Uni.EndsWith(U'Ç̌',U'Ç̌','NFC') = TRUE);
+    DATA r1 := x'43002703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'c700';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test41 := ASSERT(Uni.EndsWith(t1,t2,'NFC') = TRUE);
+    DATA r1 := x'43002703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'c700';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test42 := ASSERT(Uni.EndsWith(t1,t2,'NFD') = TRUE);
+    DATA r1 := x'43002703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'c700';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test43 := ASSERT(Uni.EndsWith(t1,t2,'NFKC') = TRUE);
+    DATA r1 := x'43002703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'c700';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test44 := ASSERT(Uni.EndsWith(t1,t2,'NFKD') = TRUE);
+    DATA r1 := x'43000C032703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'0C032703';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test45 := ASSERT(Uni.EndsWith(t1,t2,'NFC') = FALSE);
+    DATA r1 := x'43000C032703';
+    UNICODE t1 := TRANSFER(r1, UNICODE);
+    DATA r2 := x'0C032703';
+    UNICODE t2 := TRANSFER(r2, UNICODE);
+    EXPORT Test46 := ASSERT(Uni.EndsWith(t1,t2,'NFD') = TRUE);
+  END;
+
+END;

--- a/plugins/unicodelib/unicodelib.cpp
+++ b/plugins/unicodelib/unicodelib.cpp
@@ -86,6 +86,7 @@ static const char * EclDefinition =
 "  unicode UnicodeLocaleExcludeLastWord(const unicode text, const varstring localename) : c,pure,entrypoint='ulUnicodeLocaleExcludeLastWord';\n"
 "  unicode UnicodeLocaleTranslate(const unicode text, unicode sear, unicode repl) :c,pure,entrypoint='ulUnicodeLocaleTranslate';\n"
 "  boolean UnicodeLocaleStartsWith(const unicode src, unicode pref, string form) :c,pure,entrypoint='ulUnicodeLocaleStartsWith';\n"
+"  boolean UnicodeLocaleEndsWith(const unicode src, const unicode suff, const string form) :c,pure,entrypoint='ulUnicodeLocaleEndsWith';\n"
 "END;\n";
 
 static const char * compatibleVersions[] = {
@@ -747,7 +748,20 @@ void normalizationFormCheck(UnicodeString & source, const char * form)
     }
 }
 
-bool startsWith(UnicodeString & processed, UnicodeString & prefix)
+static bool endsWith(UnicodeString const & processed, UnicodeString const & suffix)
+{
+    if (processed.isEmpty() || suffix.isEmpty())
+    {
+        return false;
+    }
+    if (!processed.endsWith(suffix))
+    {
+        return false;
+    }
+    return true;
+}
+
+static bool startsWith(UnicodeString & processed, UnicodeString & prefix)
 {
     if (processed.isEmpty() || prefix.isEmpty())
     {
@@ -1642,4 +1656,17 @@ UNICODELIB_API bool UNICODELIB_CALL ulUnicodeLocaleStartsWith(unsigned srcLen, U
         normalizationFormCheck(pre, form);
     }
     return startsWith(pro, pre);
+}
+
+UNICODELIB_API bool UNICODELIB_CALL ulUnicodeLocaleEndsWith(unsigned srcLen, UChar const * src, unsigned suffLen, UChar const * suff, unsigned formLen, char const * form)
+{
+    UnicodeString pro(src, srcLen);
+    UnicodeString suf(suff, suffLen);
+    suf.trim();
+    if (formLen == 3 || formLen == 4)
+    {
+        normalizationFormCheck(pro, form);
+        normalizationFormCheck(suf, form);
+    }
+    return endsWith(pro, suf);
 }

--- a/plugins/unicodelib/unicodelib.hpp
+++ b/plugins/unicodelib/unicodelib.hpp
@@ -105,6 +105,7 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleExcludeNthWord(unsigned & tgt
 UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleExcludeLastWord(unsigned & tgtLen, UChar * & tgt, unsigned textLen, UChar const * text, char const * localename);
 UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleTranslate(unsigned & tgtLen, UChar * & tgt, unsigned textLen, UChar const * text, unsigned searLen, UChar const * sear, unsigned replLen, UChar * repl);
 UNICODELIB_API bool UNICODELIB_CALL ulUnicodeLocaleStartsWith(unsigned srcLen, UChar const * src, unsigned prefLen, UChar const * pref, unsigned formLen, char const * form);
+UNICODELIB_API bool UNICODELIB_CALL ulUnicodeLocaleEndsWith(unsigned srcLen, UChar const * src, unsigned suffLen, UChar const * suff, unsigned formLen, char const * form);
 }
 
 #endif

